hibike/hibike_process.py

```diff
@@ -85,25 +85,123 @@ def recv_subscription_requests(conn, uid_queue, stop_event):
-        event_list.append(curr_event)
-    for thread in thread_list:
-        thread.start()
-    for (index, reader) in enumerate(read_queues):
+        maybe_device = namedtuple("MaybeDevice", ["serial_conn", "queue", "event", "thread"])
+        maybe_device.queue = queue.Queue()
+        maybe_device.event = threading.Event()
+        maybe_device.serial_conn = conn
+        maybe_device.thread = threading.Thread(target=recv_subscription_response,
+                                               args=(conn, maybe_device.queue, maybe_device.event))
+        candidates.append(maybe_device)
+    for cand in candidates:
+        cand.thread.start()
+    for cand in candidates:
         try:
-            uid = reader.get(block=True, timeout=IDENTIFY_TIMEOUT)
-            device_map[serial_conns[index].name] = uid
+            uid = cand.queue.get(block=True, timeout=IDENTIFY_TIMEOUT)
+            device_map[cand.serial_conn.name] = uid
+            # Shut device up
+            hm.send(cand.serial_conn, hm.make_subscription_request(uid, [], 0))
         except queue.Empty:
             pass
-    for (event, thread) in zip(event_list, thread_list):
-        event.set()
-        thread.join()
+    for cand in candidates:
+        cand.event.set()
+        cand.thread.join()
     return device_map
 
+
+def spin_up_device(serial_port, uid, state_queue, batched_data, error_queue):
+    """
+    Spin up a device with a given UID on SERIAL_PORT.
+
+    Returns:
+        The new device.
+    """
+    pack = namedtuple("Threadpack", ["read_thread", "write_thread",
+                                     "write_queue", "serial_port"])
+    pack.write_queue = queue.Queue()
+    pack.serial_port = serial_port
+    pack.write_thread = threading.Thread(target=device_write_thread,
+                                         args=(serial_port, pack.write_queue))
+    pack.read_thread = threading.Thread(target=device_read_thread,
+                                        args=(uid, serial_port, pack.write_queue, error_queue,
+                                              state_queue, batched_data))
+    pack.write_thread.start()
+    pack.read_thread.start()
+    return pack
+
+
+def hotplug(devices, state_queue, batched_data, error_queue):
+    """
+    Remove disconnected devices and scan for new ones.
+    """
+    clean_up_queue = queue.Queue()
+    clean_up_thread = threading.Thread(target=clean_up_devices, args=(clean_up_queue, ))
+    clean_up_thread.start()
+    while True:
+        time.sleep(HOTPLUG_POLL_INTERVAL)
+        scan_for_new_devices(devices, state_queue, batched_data, error_queue)
+        remove_disconnected_devices(error_queue, devices, clean_up_queue, state_queue)
+
+
+def scan_for_new_devices(existing_devices, state_queue, batched_data, error_queue):
+    """
+    Scan for devices not in EXISTING_DEVICES, and add any if found.
+    """
+    ports, names = get_working_serial_ports(map(lambda d: d.serial_port.name,
```

Comment:
Opening a serial port multiple times is not an issue. The issue is two instances issuing instructions at the same time.